src/test/java/io/gravitee/policy/oauth2/Oauth2PolicyV4EmulationEngineCnfIntegrationTest.java

```diff
@@ -0,0 +1,483 @@
+                        .putHeader("Authorization", "Bearer " + DummyOAuth2Resource.TOKEN_SUCCESS_WITH_CLIENT_ID)
+                        .putHeader("ssl-client-cert", encoded)
+                        .rxSend()
+                );
+
+            assert401unauthorized(wiremock, httpClientResponse);
+        }
+
+        protected OngoingStubbing<Optional<Subscription>> whenSearchingSubscription(String api, String clientId, String plan) {
+            return when(getBean(SubscriptionService.class).getByApiAndSecurityToken(eq(api), securityTokenMatcher(clientId), eq(plan)));
+        }
+    }
+
+    @Nested
+    @DeployApi("/apis/oauth2.json")
+    @GatewayTest
+    public class Oauth2PolicyV4EmulationEngineCnfPeerCertificateIntegrationTest
+        extends AbstractOauth2PolicyCnfPeerCertificateIntegrationTest {}
+
+    public static class AbstractOauth2PolicyCnfPeerCertificateIntegrationTest
+        extends AbstractPolicyTest<Oauth2Policy, OAuth2PolicyConfiguration> {
+
+        @Override
+        public void configureGateway(GatewayConfigurationBuilder gatewayConfigurationBuilder) {
+            Oauth2PolicyV4EmulationEngineCnfIntegrationTest.configureGateway(gatewayConfigurationBuilder);
+        }
+
+        @Override
+        protected void configureHttpClient(final HttpClientOptions options) {
+            Oauth2PolicyV4EmulationEngineCnfIntegrationTest.configureHttpClient(options, gatewayPort());
+
+            final PemKeyCertOptions pemKeyCertOptions = new PemKeyCertOptions();
+            pemKeyCertOptions.setCertPath(Oauth2PolicyV3IntegrationTest.class.getResource("/certificate/client1-crt.pem").getPath());
+            pemKeyCertOptions.setKeyPath(Oauth2PolicyV3IntegrationTest.class.getResource("/certificate/client1-key.pem").getPath());
+            options.setPemKeyCertOptions(pemKeyCertOptions);
+        }
+
+        @Override
+        public void configureApi(final Api api) {
+            Plan oauth2Plan = createOauth2Plan(api);
+
+            OAuth2PolicyConfiguration configuration = new OAuth2PolicyConfiguration();
+            configuration.setOauthResource("dummy-oauth2-resource");
+            configuration.getConfirmationMethodValidation().getCertificateBoundThumbprint().setEnabled(true);
+            try {
+                oauth2Plan.setSecurityDefinition(new ObjectMapper().writeValueAsString(configuration));
+            } catch (JsonProcessingException e) {
+                throw new RuntimeException("Failed to set OAuth2 policy configuration", e);
+            }
+
+            api.setPlans(Collections.singletonList(oauth2Plan));
+        }
+
+        @Override
+        public void configureResources(final Map<String, ResourcePlugin> resources) {
+            Oauth2PolicyV4EmulationEngineCnfIntegrationTest.configureResources(resources);
+        }
+
+        @Test
+        void should_access_api_with_valid_certificate(HttpClient client) throws InterruptedException, URISyntaxException, IOException {
```

Comment:
I didn't realize the http client was configured with a keystore in the first place. Just to make it obvious, can you rename it to `should_access_api_with_valid_certificate_from_ssl_session`?

Reply:
The class name is used to categorized the test. Here the class is peer certificate

Reply:
Ok I didn't notice that the `Peer` certificate stands for "from SSL Session". I saw other methods with the suffix "from_header" so I was confused.